typing/persistent_env.ml

```diff
@@ -210,7 +213,9 @@ let find_pers_struct ~allow_hidden penv val_of_pers_sig check name =
   let {persistent_structures; _} = penv in
   if name = "*predef*" then raise Not_found;
   match Hashtbl.find persistent_structures name with
-  | Found (ps, pm) -> (ps, pm)
+  | Found (ps, pm) when allow_hidden || ps.ps_visibility = Load_path.Visible ->
+    (ps, pm)
+  | Found _ -> raise Not_found
   | Missing -> raise Not_found
```

Comment:
I think the logic around `Missing` isn't quite right. If you first try to load a file with `allow_hidden` is `false` then you'll set it to `Missing` in the cache, and that will prevent future lookups even when `allow_hidden` is `true`.

Reply:
For future reference: Leo and I discussed this offline.  It should be difficult to get bad behavior out of this, because generally if `~allow_hidden` is false we're checking an explicit reference to something in a program, and a failure here will result in a failure of typechecking.

We're not confident that's always the case, though, so I'm pushing a fix which is to only add `Missing` to the cache if `allow_hidden`.

An alternative would be to add an argument to `Missing` to track which kind of lookup has failed.  That would have the advantage of avoiding multiple calls to `can_load_cmis`, but only in the rare case that typechecking succeeds after multiple failing calls to `find_pers_struct` for the same module with `allow_hidden=false`.  (And the disadvantage of marginal additional complexity and allocation).

Reply:
The only case that I can think of as potentially problematic would be in Merlin, if they cache failed lookups usefully. (But this assume they watch the filesystem somehow to know when to clean the cache.) Do we know if they do that? Probably not?

(I would be slightly more at ease with caching both with an extra argument to `Missing`, which we are sure will have no adverse side-effect.)

Reply:
IIRC they already have to patch this stuff to get their desired behaviour anyway